sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -61,68 +52,52 @@ final class Decimal extends Ordered[Decimal] with Serializable {
    * Set this Decimal to the given Int. Will have precision 10 and scale 0.
    */
   def set(intVal: Int): Decimal = {
-    this.decimalVal = null
-    this.longVal = intVal
-    this._precision = 10
-    this._scale = 0
+    decimalVal = JavaBigDecimal.valueOf(intVal)
+    _precision = 10
     this
   }
 
   /**
    * Set this Decimal to the given unscaled Long, with a given precision and scale.
+   *
+   * Note: this is used in serialization, caller will make sure that it will not overflow
    */
   def set(unscaled: Long, precision: Int, scale: Int): Decimal = {
-    if (setOrNull(unscaled, precision, scale) == null) {
```

Comment:
Do you think that we should replace this old check with an assert, just to be safe / guard against mistakes made by the caller?